jax/_src/lax/control_flow/loops.py

```diff
@@ -797,58 +1111,84 @@ def _scan_dce_rule(used_outputs: List[bool], eqn: core.JaxprEqn
       used_carry_out = _map(operator.or_, used_carry_out, used_carry_in)
   else:
     assert False, "Fixpoint not reached"
-  if config.jax_enable_checks: core.check_jaxpr(jaxpr.jaxpr)
+  if config.jax_enable_checks: core.check_jaxpr(jaxpr_dce)
 
-  new_linear = [l for l, u in zip(eqn.params['linear'], used_inputs) if u]
+  start_stop_done_linear, rest_linear = split_list(eqn.params["linear"], [3])
+  new_linear = [l for l, u in zip(rest_linear, used_inputs) if u]
+  new_linear = start_stop_done_linear + new_linear
   new_params = dict(eqn.params, num_consts=sum(used_consts),
                     num_carry=sum(used_carry_in), linear=tuple(new_linear),
                     jaxpr=core.ClosedJaxpr(jaxpr_dce, jaxpr.consts))
   # TODO(mattjj,sharadmv): don't assume effects are never DCE'd?
   new_eqn = pe.new_jaxpr_eqn(
-      [v for v, used in zip(eqn.invars, used_inputs) if used],
-      [v for v, used in zip(eqn.outvars, used_outputs) if used],
+      [v for v, used in zip(eqn.invars, [True, True, True] + used_inputs) if used],
+      [v for v, used in zip(eqn.outvars, [True, True] + used_outputs) if used],
       eqn.primitive, new_params, eqn.effects, eqn.source_info)
-  assert len(new_eqn.invars ) == len(new_params['jaxpr'].in_avals )
-  assert len(new_eqn.outvars) == len(new_params['jaxpr'].out_avals)
-  return used_inputs, new_eqn
+  assert len(new_eqn.invars ) == 3 + len(new_params['jaxpr'].in_avals )
+  assert len(new_eqn.outvars) == 2 + len(new_params['jaxpr'].out_avals)
+  return [True, True, True] + used_inputs, new_eqn
 
 # TODO(mattjj): de-duplicate code with _scan_partial_eval
 def _scan_partial_eval_custom(saveable, unks_in, inst_in, eqn):
   jaxpr = eqn.params['jaxpr']
   num_consts, num_carry = eqn.params['num_consts'], eqn.params['num_carry']
-  num_ys = len(jaxpr.out_avals) - num_carry
+  num_ys = len(jaxpr.out_avals) - 1 - num_carry
+
+  [start_uk, stop_uk, done_uk], const_uk, carry_uk, xs_uk = split_list(unks_in, [3, num_consts, num_carry])
+  if done_uk:
+    # If conditional is unknown, just do the default processing.
+    eqn_known = None
+    eqn_staged = eqn
+    unks_out = [True] * len(eqn.outvars)
+    inst_out = [True] * len(eqn.outvars)
+    new_vars = [x for inst, x in zip(inst_in, eqn.invars)
+                if type(x) is core.Var and not inst]
+    return eqn_known, eqn_staged, unks_out, inst_out, new_vars
```

Comment:
I don't think there's a hook for this default processing, so I essentially just ended up copying what `pe.partial_eval_jaxpr_custom` does by default.